dask/array/optimization.py

```diff
@@ -290,3 +299,182 @@ def fuse_slice(a, b):
+
+    Given a set of TOP layers, combine them into a single layer.  The provided
+    layers are expected to fit well together.  That job is handled by
+    ``optimize_atop``
+
+    Parameters
+    ----------
+    inputs : List[TOP]
+
+    Returns
+    -------
+    top : TOP
+
+    See Also
+    --------
+    optimize_atop
+    """
+    from dask.core import reverse_dict
+    from dask.array.core import TOP
+    inputs = {inp.output: inp for inp in inputs}
+    dependencies = {inp.output: {d for d, v in inp.indices if v is not None and d in inputs}
+                    for inp in inputs.values()}
+    dependents = reverse_dict(dependencies)
+
+    new_index_iter = iter('ABCDEFGHIJKLMNOPQRSTUVWXYZ')
+
+    [root] = [k for k, v in dependents.items() if not v]
+
+    indices = list(inputs[root].indices)
+    new_axes = inputs[root].new_axes
+    concatenate = inputs[root].concatenate
+
+    dsk = dict(inputs[root].dsk)
+
+    seen = set()
+
+    changed = True
+    while changed:
+        changed = False
+        for i, (dep, ind) in enumerate(indices):
+            if ind is None:
+                continue
+            if dep not in inputs:
+                continue
+
+            changed = True
+
+            # Replace _n with dep name in existing tasks
+            # (inc, _0) -> (inc, 'b')
+            dsk = {k: subs(v, {atop_token(i): dep}) for k, v in dsk.items()}
+
+            # Remove current input from input indices
+            # [('a', 'i'), ('b', 'i')] -> [('a', 'i')]
+            _, current_dep_indices = indices.pop(i)
+            sub = {atop_token(i): atop_token(i - 1) for i in range(i + 1, len(indices) + 1)}
+            dsk = subs(dsk, sub)
+
+            # Change new input_indices to match give index from current computation
+            # [('c', j')] -> [('c', 'i')]
+            new_indices = inputs[dep].indices
+            sub = dict(zip(inputs[dep].output_indices, current_dep_indices))
+            contracted = {x for _, j in new_indices if j is not None for x in j if x not in inputs[dep].output_indices}
+            extra = dict(zip(contracted, new_index_iter))
+            sub.update(extra)
+            new_indices = [(x, index_subs(j, sub)) for x, j in new_indices]
+
+            # TODO: handle new_axes
+
+            # Bump new inputs up in list
+            sub = {}
+            for i, index in enumerate(new_indices):
+                if index not in indices:  # use old inputs if available
+                    sub[atop_token(i)] = atop_token(len(indices))
+                    indices.append(index)
+                else:
+                    sub[atop_token(i)] = atop_token(indices.index(index))
+            new_dsk = subs(inputs[dep].dsk, sub)
+
+            # indices.extend(new_indices)
+            dsk.update(new_dsk)
+
+    indices = [(a, tuple(b) if isinstance(b, list) else b) for a, b in indices]
+
+    # De-duplicate indices
+    new_indices = []
+    seen = dict()
+    sub = dict()  # like {_0: _0, _1: _0, _2: _1}
```

Comment:
Nit: it's faster, and more readable (IMO) to use `{}` instead of `dict()`.

Reply:
Done (although I slightly disagree about readability.  `dict()` is maybe more clearly an empty dict).